core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -539,7 +539,7 @@ private[deploy] class Master(
 
   private def completeRecovery() {
     // Ensure "only-once" recovery semantics using a short synchronization period.
-    if (state != RecoveryState.RECOVERING) { return }
+    if (state != RecoveryState.RECOVERING && state != RecoveryState.COMPLETING_RECOVERY) { return }
```

Comment:
It seems better to close the master as you say if an exception happened during recovery.So I change the last commit .